include/fiction/algorithms/network_transformation/inverter_balancing.hpp

```diff
@@ -0,0 +1,320 @@
+                }
+                if constexpr (mockturtle::has_is_nary_or_v<TopoNtkSrc> && mockturtle::has_create_nary_or_v<Ntk>)
+                {
+                    if (ntk.is_nary_or(g))
+                    {
+                        old2new[g] = ntk_dest.create_nary_or(children);
+                        return true;
+                    }
+                }
+                if constexpr (mockturtle::has_is_nary_xor_v<TopoNtkSrc> && mockturtle::has_create_nary_xor_v<Ntk>)
+                {
+                    if (ntk.is_nary_xor(g))
+                    {
+                        old2new[g] = ntk_dest.create_nary_xor(children);
+                        return true;
+                    }
+                }
+                if constexpr (mockturtle::has_node_function_v<TopoNtkSrc> && mockturtle::has_create_node_v<Ntk>)
+                {
+                    old2new[g] = ntk_dest.create_node(children, ntk.node_function(g));
+                    return true;
+                }
+
+                return true;
+            });
+
+        ntk.foreach_po(
+            [this, &ntk_dest, &old2new](const auto& po)
+            {
+                auto tgt_signal = old2new[ntk.get_node(po)];
+                auto tgt_po     = ntk.is_complemented(po) ? ntk_dest.create_not(tgt_signal) : tgt_signal;
+                if (tgt_po == 0)
+                {
+                    tgt_signal = old2new[ntk.get_node(preserved_po[0])];
+                    preserved_po.erase(preserved_po.begin());
+                    tgt_po = ntk.is_complemented(po) ? ntk_dest.create_not(tgt_signal) : tgt_signal;
+                }
+                ntk_dest.create_po(tgt_po);
+            });
+
+        if constexpr (mockturtle::has_foreach_ri_v<TopoNtkSrc>)
+        {
+            ntk.foreach_ri(
+                [this, &ntk_dest, &old2new](const auto& po)
+                {
+                    const auto tgt_signal = old2new[ntk.get_node(po)];
+                    const auto tgt_po     = ntk.is_complemented(po) ? ntk_dest.create_not(tgt_signal) : tgt_signal;
+
+                    ntk_dest.create_ri(tgt_po);
+                });
+        }
+
+        // restore signal names if applicable
+        fiction::restore_names(ntk, ntk_dest, old2new);
+
+        return ntk_dest;
+    }
+
+  private:
+    using TopoNtkSrc = mockturtle::topo_view<Ntk>;
+    TopoNtkSrc                          ntk;
+    mockturtle::fanout_view<TopoNtkSrc> fo_ntk{ntk};
+    std::vector<mockturtle::node<Ntk>>  x_inv{};         // inverter nodes, which get deleted
+    std::vector<mockturtle::node<Ntk>>  m_inv{};         // inverter nodes, which get moved to fanin position
+    std::vector<mockturtle::node<Ntk>>  blc_fos{};       // fo nodes, where balancing is applied
+    std::vector<mockturtle::node<Ntk>>  preserved_po{};  // nodes where pos need to be preserved
+};
+
+template <typename NtkSrc>
+bool inverter_balancing_recursive(const NtkSrc& ntk)
+{
+    bool       return_val = false;
```

Comment:
What does `return_val` mean? It must represent something, no? Can we adjust the name accordingly?